core/store/src/trie/mod.rs

```diff
@@ -550,27 +553,48 @@ impl WrappedTrieChanges {
         self.trie_changes.deletions_into(self.trie.clone(), store_update)
     }
 
-    pub fn key_value_changes_into(
-        &self,
+    /// Save state changes into Store.
+    ///
+    /// NOTE: the changes are drained from `self`.
+    pub fn state_changes_into(
+        &mut self,
         store_update: &mut StoreUpdate,
     ) -> Result<(), Box<dyn std::error::Error>> {
         store_update.trie = Some(self.trie.clone());
-        for (key, changes) in &self.kv_changes {
+        for (key, raw_changes) in
+            std::mem::replace(&mut self.state_changes, Default::default()).into_iter()
```

Comment:
`std::mem::replace(..., Default::default())` can be replaced with `std::mem::take()` in stable Rust for a few releases, but we have an old nightly. I would love to have us following Rust stable releases, so our nightly has all the stable features.

Also, Rust 1.42 `matches!` macro can be used in the `assert!` below.